src/web/handlers.rs

```diff
@@ -88,36 +136,75 @@ pub async fn get_tiles(
-        }
-        Err(e) => {
-            match e.kind() {
-                HandlerErrorKind::BadAdmResponse(es) => {
-                    warn!("Bad response from ADM: {:?}", e);
-                    metrics.incr_with_tags("tiles.invalid", Some(&tags));
-                    // Report directly to sentry
-                    // (This is starting to become a pattern. 🤔)
-                    let mut tags = Tags::from_head(request.head(), settings);
-                    tags.add_extra("err", es);
-                    tags.add_tag("level", "warning");
-                    l_sentry::report(&tags, sentry::event_from_error(&e));
-                    warn!("ADM Server error: {:?}", e);
-                    Ok(HttpResponse::NoContent().finish())
+    let handle_result = || {
+        match result {
+            Ok(response) => {
+                let tiles = cache::Tiles::new(response, add_jitter(&state.settings))?;
+                trace!(
+                    "get_tiles: cache miss{}: {:?}",
+                    if expired { " (expired)" } else { "" },
+                    &audience_key
+                );
+                metrics.incr("tiles_cache.miss");
+                state.tiles_cache.insert(
+                    audience_key.clone(),
+                    TilesState::Fresh {
+                        tiles: tiles.clone(),
+                    },
+                );
+                Ok(content_response(&tiles.content))
+            }
+            Err(e) => {
+                // Add some kind of stats to Retrieving or RetrievingFirst?
+                // do we need a kill switch if we're restricting like this already?
+                match e.kind() {
+                    HandlerErrorKind::BadAdmResponse(es) => {
+                        warn!("Bad response from ADM: {:?}", e);
+                        metrics.incr_with_tags("tiles.invalid", Some(&tags));
+                        state.tiles_cache.insert(
+                            audience_key.clone(),
+                            TilesState::Fresh {
+                                tiles: Tiles::empty(add_jitter(&state.settings)),
+                            },
+                        );
+                        // Report directly to sentry
+                        // (This is starting to become a pattern. 🤔)
+                        let mut tags = Tags::from_head(request.head(), settings);
+                        tags.add_extra("err", es);
+                        tags.add_tag("level", "warning");
+                        l_sentry::report(&tags, sentry::event_from_error(&e));
+                        warn!("ADM Server error: {:?}", e);
+                        Ok(HttpResponse::NoContent().finish())
+                    }
+                    _ => Err(e),
                 }
-                _ => Err(e),
             }
         }
+    };
+
+    let result = handle_result();
+    // Cleanup the TilesState on errors
+    // TODO: potential panics are not currently cleaned up
```

Comment:
I'm hoping we could move this cleanup section into a `Drop` impl of some new struct to handle this. The struct could probably also handle inserting/altering the new state values above and hopefully simplfy the handler a bit